server/app/services/cloud/websocket_client.rb

```diff
@@ -124,55 +135,112 @@ def send_message(msg)
     def on_open(event)
       ping_timer.cancel if ping_timer
       info "cloud connection opened to #{self.api_uri}"
+      subscribe_events(EventStream.channel)
       @connected = true
       @connecting = false
     end
 
     # @param [Faye::WebSocket::API::Event] event
     def on_message(event)
+      debug "Received websocket message"
+      if leader?
         data = MessagePack.unpack(event.data.pack('c*'))
-        if request_message?(data)
-          EM.defer {
-            if leader?
-              response = rpc_server.handle_request(data)
-              send_message(MessagePack.dump(response).bytes)
-            end
-          }
-        elsif notification_message?(data)
-          EM.defer {
-            if leader?
-              rpc_server.handle_notification(data)
-            end
-          }
-        end
+        EM.defer {
+          if request_message?(data)
+            debug "Creating RPC request"
+            response = rpc_server.handle_request(data)
+            send_message(MessagePack.dump(response).bytes)
+          elsif notification_message?(data)
+            rpc_server.handle_notification(data)
+          end
+        }
+      else
+        debug "Ignoring request because not leader"
+      end
     rescue => exc
       error exc.message
     end
 
     # @param [Faye::WebSocket::API::Event] event
     def on_close(event)
+      @ping_timer = nil
+      @close_timer.cancel if @close_timer
+      @close_timer = nil
       @connected = false
       @connecting = false
       @ws = nil
       if event.code == 1002
         handle_invalid_token
       end
       info "cloud connection closed with code: #{event.code}"
+      unsubscribe_events
     rescue => exc
       error exc.message
     end
 
+    # @param [Hash] msg
+    def send_notification_message(msg)
+      invalidate_users_cache if msg[:type] == 'User' # clear cache if users are modified
+      grid_id = resolve_grid_id(msg)
+      users = resolve_users(grid_id)      
+      params = [grid_id, users, msg[:object]]
+      message = [2, "#{msg[:type]}##{msg[:event]}", params]
```

Comment:
Could these use similar `/grid_service/update` style methods as the Agent RPC does? Ideally they would be something vaugely consistent with the HTTP API URL structure...